DOGM Display Screen Sleep

borland1 opened the pull request:
### Description

Adds Screen Sleep feature for U8g DOGM displays which supports Sleep and Wake functions.
  - #define SCREEN_TIMEOUT in `Configuration.h` file is, by default, commented out (disabled).
  - Adds to LCD Menu under Configuration Menu, a timeout setting menu.
  - User can disable screen timeout feature by setting timeout value to '0'.
  - Timeout setting changes will be saved to EEPROM with other setting changes being saved.
  - Supports timeout times between 1 and 99 minutes in 1 minute increments. 
  - Rotation of rotary encoder, or press of rotary encoder button, wakes display when in display is in sleep mode. 


### Requirements

U8g DOGM supported displays.

### Benefits

Helps reduce energy consumption during printing.  Reduces pixel burn-in with OLED displays.

### Configurations

`Configuration.h` file has been modified and is included with this PR.

### Related Issues

None found.

borland1 committed:
Update Configuration.h

Add Screen Timeout Feature for DOGM displays.

borland1 committed:
Update Conditionals_LCD.h

Adding #define HAS_SCREEN_TIMEOUT 1

borland1 committed:
Update language_en.h

Add to list: LSTR MSG_SCREEN_TIMEOUT  = _UxGT("ScreenTimeout(m)");

borland1 committed:
Update settings.cpp

Add settings for SCREEN_TIMEOUT feature.

borland1 committed:
Update marlinui.h

Add declarations for SCREEN_TIMEOUT feature.

borland1 committed:
Update marlinui.cpp

Add SCREEN_TIMEOUT feature

borland1 committed:
Update menu_configuration.cpp

Add menu EDIT_ITEM: SCREEN_TIMEOUT

borland1 committed:
Update marlinui_DOGM.cpp

Add  ui.sleep_on() and ui.sleep_off() for SCREEN_TIMEOUT feature.

thisiskeithb commented:
This would be good, but I think it should be extra clear in the config/header that this won't work on the most common 12864 displays (RepRapDiscount Full Graphic Smart Controller & Mini 12864 LCDs) because I can already see the support requests coming in asking why it doesn't work.

Adding some sanity checks to limit use would probably also help prevent *hopeful* LCD configs.

borland1 commented:
Thanks for the good comments Keith,   I don't know all the displays supported by Marlin, or the number of sales with each, but can, like you suggested, add some sanity checks in the `SanityChecks.h` file for a few of the displays listed in the `Configuration.h` file. 

This PR's code does rely on or interfere with any code related to the TOUCH_IDLE_SLEEP or LCD_BACKLIGHT_TIMEOUT features currently in bugfix-2.0.x, and those are available to anyone with those type of displays needing those features.

I did place the `#define`enabling line just below the _OLED Displays_ heading, so it is somewhat hidden.  I did write a lengthy paragraph explaining that the feature is only for displays that meet a specific hardware requirement, but needed to also explain functionality too, so I don't think it's appropriate to add more words there.  What I provided in the comments should be sufficient.

luxflow commented:
I think adding sanity check for most commonly(?) used display (12864, 2004, mini 12864) which are supported by other firmware(such as [klipper](https://github.com/Klipper3d/klipper/blob/master/config/sample-lcd.cfg)) will be sufficient  

if filtering those display with driver type
https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/lcd/dogm/marlinui_DOGM.h

- HD44780 based display (reprapdiscount smartcontroller 2004 lcd)
  - HAS_MARLINUI_HD44780
- st7920 based display (reprapdiscount fullgraphic smartcontroller 12864 lcd)
  - REPRAPWORLD_GRAPHICAL_LCD
  - IS_U8GLIB_ST7920
- mini12864 display
  - FYSETC_MINI_12864, MKS_MINI_12864, ENDER2_STOCKDISPLAY
  - MINIPANEL

should be undefined or disabled and DOGLCD always should be defined

I think these will cover most of use case isn't it?

luxflow commented on Marlin/src/inc/Conditionals_LCD.h:
```diff
@@ -1427,3 +1427,9 @@
 #endif
+
+/**
+ * Screen Timeout Feature
```
it seems that `*/` is missing

borland1 replied:
Fixed, thanks for the help.

borland1 committed:
Update Conditionals_LCD.h

Fixed typo on multi line comments, missing closing "*/" characters.

borland1 committed:
Create SanityCheck.h

Merge with FR #23896  Sanity Checks for Screen Support Feature,  Changes adds checks for unsupported displays.

luxflow commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -2873,6 +2873,29 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
+    #error "SCREEN_TIMEOUT feature not supported by Reprap Discount Full Graphics Smart Controller using ST7920"
+  #elif ENABLED(CARTESIO_UI)
+    #error "SCREEN_TIMEOUT feature not supported by CARTESIO_UI LCD"
+  #elif IS_U8GLIB_LM6059_AF
```
CARTESIO_UI, IS_U8GLIB_LM6059_AF, IS_U8GLIB_ST7565_64128 devices support sleep
https://github.com/MarlinFirmware/U8glib-HAL/blob/master/src/clib/u8g_dev_st7565_dogm128.c#L171-L176
https://github.com/MarlinFirmware/U8glib-HAL/blob/master/src/clib/u8g_dev_st7565_lm6059.c#L144-L149
https://github.com/MarlinFirmware/U8glib-HAL/blob/master/src/clib/u8g_dev_st7565_64128n.c#L185-L190

thinkyhead replied:
Should I add these displays to the updates at borland1/Marlin#1 ?

luxflow replied:
it is ok to add

luxflow replied:
By the way, If you don't mind, could you merge u8glib-hal for supporting additional devices?

luxflow commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -2873,6 +2873,29 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
+*/
+#if HAS_SCREEN_TIMEOUT
+  #if HAS_MARLINUI_HD44780
+    #error "SCREEN_TIMEOUT feature not supported by HD44780 Reprap Discount Smart Controller 2004 LCD"
```
Rather than checking HD44780 only
checking`HAS_MARLINUI_U8GLIB` seems better option

luxflow commented:
I PR to add some display sleep support for u8glib-hal which is backported from u8g2
After this PR is merged, we can add support for ST7920 and MINI12864 based display which are quite commonly used (reprap full graphic smart controller, mini12864, minipanel)

thinkyhead commented:
For some reason I'm unable to add commits directly to this PR so please merge borland1/Marlin#1 to apply my updates.

borland1 committed:
Update SanityCheck.h

In Screen Timeout section, delete sanity checks for 'HAS_MARLINUI_HD44780' and  'CARTESIO_UI'.  Reason: Screen Timeout feature is for DOGM displays, but HD44780 is not DOGM, while CARTESIO_UI displays are supported by Screen Timeout feature.   Merge changes with PR #23927

thinkyhead commented:
All you need to do to merge my PR into this PR is to press the "Merge" button on the page borland1/Marlin#1.

borland1 commented:
Sorry for the untimely response, but I have not had time to do the merge because there are some problems with the changes you are proposing.   Right now, I don't have spare time to work on a solution that implements a compromise.  Some of your changes appear either out of scope, not well thought out, or are trivial.   I should have something worked up to merge in three or four days.

thinkyhead commented:
> Some of your changes appear either out of scope, not well thought out, or are trivial.

I've reviewed the changes in my cleanup and extension PR, and I consider them all pretty much mandatory. I would prefer that you merge my PR and then afterward apply your adjustments.
